### ISSUE TITLE
Write and Test - Truffle Tests for Smart Contracts, on Forked Mainnet

### ISSUE THREAD

**Issue by amateur-dev** [amateur-dev/Unipool#1]
Objective: This Bounty is for stress testing the [zUniPool Contract](https://github.com/amateur-dev/Unipool/blob/master/contracts/zUniPool.sol), using, Truffle test and Forked Mainnet.

Please go through the [ReadMe](https://github.com/amateur-dev/Unipool/blob/master/README.md) in detail.  

Once you have written the complete Test Script, please submit it as a PR. 

**Scope of the test:** The Scope of the test should cover, the following, **at least**:

1. First user is able to stake LP tokens at a price of 1:1
2. Second user is also able to stake LP, immediately, at a price of 1:1
3. Immediately, the First User is able to withdraw its LP with the same number of LP tokens
4. Immediately, the Second User is able to check the value of its zUNI tokens and it should return, at least, the same number of LP tokens that it staked (or higher)
5. Third User is able to stake LP tokens,  *after a week of time lapse*, but the zUNI Tokens issued to the Third User is not at a price ratio of 1:1, but lesser number of zUNI Tokens compared to the number of LP tokens staked
6. At the same time, Second User wants to burn [ `getMyStakeOut` ] a exactly half of its zUNI Tokens, and it is able to receive LP tokens.
    - The LP tokens received at this stage by Second User is *higher* than *half* of the LP Tokens staked by the Second User

**Goal**: One should be able to run `truffle test test/name_of_the_script`, testing on a ganache testnet, which is run by using the `ganache-cli —fork`

**Note**:  For the purpose of this test, you will be required to acquire LP tokens.  You could do so, {in the forked mainnnet}, by calling the function `LetsInvest(address _TokenContractAddress, address _towhomtoissue)` in the [UnipoolGeneralZap Smart Contract](https://github.com/amateur-dev/Unipool/blob/master/contracts/Unipool_GeneralZap_Flat.sol), running on the mainnet at `0x97402249515994Cc0D22092D3375033Ad0ea438A`.

Additional Token Information, for ease of reference:

sETH Token Address: `0x5e74c9036fb86bd7ecdcb084a0673efc32ea31cb`

sETH_LP_Address: `0xe9Cf7887b93150D4F2Da7dFc6D502B216438F244`

Original UniPool: `0x48D7f315feDcaD332F68aafa017c7C158BC54760` {the contract developed and deployed by Synthetix Team}

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 400.0 DAI (400.0 USD @ $1.0/DAI)  attached to it as part of the @DeFiStrategies fund.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/amateur-dev/Unipool/1/4143).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/amateur-dev/Unipool/1/4143).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $78,925.15 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__Work has been started__.


These users each claimed they can complete the work by 1 year, 11 months ago.
Please review their action plans below:


**1) [pbteja1998](https://gitcoin.co/pbteja1998) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/amateur-dev/unipool/1/4143?mutate_worker_action=approve&worker=pbteja1998) | [reject worker](https://gitcoin.co/issue/amateur-dev/unipool/1/4143?mutate_worker_action=reject&worker=pbteja1998))_.**

I will write tests based on the requirements given. Then acquire LP tokens and then will test accordingly. I have previously done bounties related to testing smart contracts.
**2) [bobface](https://gitcoin.co/bobface) has been approved to start work.**

I have written lots of contract unit tests in the past and am familiar with liquidity pools. I have completed bounties to deploy custom forks of Uniswap in the past. This way I will be able to quickly understand the contract source and test it appropriately. Thanks for considering me!
**3) [mds1](https://gitcoin.co/mds1) has been approved to start work.**

Hey there! This seems like a straightforward task and a clear description, so I should be able to get this done for you pretty quickly.

I have a lot of experience testing against the forked mainnet—most recently with the Floatify contracts the FakerDAO contracts—so am familiar with its features and quirks. I can also add in a few events for you, since it seems the LPTokensStaked and LPTokensWithdrawn events are currently not used.
**4) [ridesolo](https://gitcoin.co/ridesolo) has been approved to start work.**

Hello, I have worked any many staking algorithms and tests units, you can check below to confirm my experience:
https://github.com/EthereumCommonwealth/Governance-System/tree/master/staking_contract
https://github.com/callisto-staking
https://github.com/staking-library
I also created a recent grant related with staking algorithm as you can see here https://gitcoin.co/grants/470/staking-library?tab=description.
Also as expert solidity auditor https//gist.github.com/ridesolo, I will try to write complete test to avoid any possible issues.
The work plan is quiet clear, so I will do my best to make it quickly.

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/amateur-dev/unipool/1/4143).

**Comment by gitcoinbot** [comment]
@iamonuwa Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/amateur-dev/Unipool/1/4143?snooze=1) | [3 days](https://gitcoin.co/issue/amateur-dev/Unipool/1/4143?snooze=3) | [5 days](https://gitcoin.co/issue/amateur-dev/Unipool/1/4143?snooze=5) | [10 days](https://gitcoin.co/issue/amateur-dev/Unipool/1/4143?snooze=10) | [100 days](https://gitcoin.co/issue/amateur-dev/Unipool/1/4143?snooze=100)

**Comment by iamonuwa** [comment]
@gitcoinbot, work in progress. Sending in PR.

**Comment by gitcoinbot** [comment]
@bobface Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/amateur-dev/Unipool/1/4143?snooze=1) | [3 days](https://gitcoin.co/issue/amateur-dev/Unipool/1/4143?snooze=3) | [5 days](https://gitcoin.co/issue/amateur-dev/Unipool/1/4143?snooze=5) | [10 days](https://gitcoin.co/issue/amateur-dev/Unipool/1/4143?snooze=10) | [100 days](https://gitcoin.co/issue/amateur-dev/Unipool/1/4143?snooze=100)

**Comment by Bobface** [comment]
@gitcoinbot I'm on it

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. **Submitted** 4. Done 

<hr>

__Work for 400.0 DAI (400.0 USD @ $1.0/DAI) has been submitted by__:

 1. [@bobface](https://gitcoin.co/bobface)



@nodar please take a look at the submitted work:
* [PR](https://github.com/amateur-dev/Unipool/pull/2) by @bobface

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/amateur-dev/Unipool/1/4143)
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/amateur-dev/Unipool/1/4143).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $80,752.21 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by RideSolo** [comment]
thanks for approving my request, I will submit the tests asap.

**Comment by Bobface** [comment]
Hey @amateur-dev I opened a PR days ago with the tests.

**Comment by mds1** [comment]
@amateur-dev @nodar Looks like a PR was submitted about a week ago, but also saw that there were two new approvals to start work—just want to double-check those approvals were intentional before starting anything. Thanks!

**Comment by RideSolo** [comment]
I think that he meant to create a "contest" project type probably, @nodar can you please confirm

**Comment by Solexplorer** [comment]
Hey @amateur-dev,
Solexplorer from Gitcoin here, I see that @Bobface has already created a PR, did you get a chance to look on it and provide him with feedback if you have any?

Looking forward to your reply

**Comment by amateur-dev** [comment]
@RideSolo apologies for the delay in responding.  I was away from desk and COVID has not been kind.  We did not intend to create a "contest" but sadly, we have did not yet get someone to give us the proper test set up in which we (a) either know what the issue is, or (b) there is no issue at all.   So we are still looking for the right PR.  cc: @mds1

**Comment by RideSolo** [comment]
@amateur-dev no problem we can all understand the situation, just let us know if you don't solve it with Bobface, so we start working on it (me or mds1)

**Comment by nodar** [comment]
Hey @Bobface let us know if you are still interested in helping out on this. Thank you for checking in guys, just noticed these. @RideSolo @mds1

**Comment by Bobface** [comment]
@nodar Yes, I wrote a long explanation on my PR. Currently waiting for response.

**Comment by amateur-dev** [comment]
Sorry @Bobface I keep replying to those emails and it seems none of them get posted over here.  Yes @Bobface please pass that test (due to the really small difference), {may be we can put a small threshold/range}.  The key test is the advanceTime. We are facing issues in doing the test for a week ahead in time, and then more time after that too (eg 1 week, 2 weeks, 4 weeks [that is it]).

**Comment by Bobface** [comment]
@amateur-dev I think on Github you can't reply to emails, you have to write your answer directly. Okay, I will implement a range for this situation.

**Comment by Bobface** [comment]
@amateur-dev The problem is that on forked mainnet you cannot use `advanceTime` because it relies on the blocks from mainnet. However there is natural time advancement occuring for every block. To be able to manually set the time you have to run the tests on a "normal" ganache testnet.  I have updated my tests to account for the differencess now and they are working - please have a look.

**Comment by amateur-dev** [comment]
Cool. Will take a look at it

**Comment by mds1** [comment]
@Bobface I believe you should be able to advance time on the forked mainnet—[here](https://github.com/ScopeLift/fakerdao/blob/master/test/faker-time.js) is an example project I did recently which successfully uses OpenZeppelin's `time.increase()` on a forked Kovan instance of ganache-cli.

I'd expect forked Kovan and mainnet instances to behave similarly in regards to advancing time, but it looks like that's what you [tried](https://github.com/amateur-dev/Unipool/pull/2/files#diff-f2294d096cc10481b27c7c7ef4b11eb3R62) in the PR, so I could be wrong.

**Comment by Bobface** [comment]
@mds1 Yup I tried it in the tests but it does not work. I left it in there in case someone wants to try themself. With natural time progression the test works, by using OpenZeppelin's `time.increase()` it fails.

**Comment by mds1** [comment]
@Bobface Huh, that is strange. If you and @amateur-dev don't mind could you create an issue for that in the [ganache-core](https://github.com/trufflesuite/ganache-core) repo? The fork feature is awesome but quite [buggy](https://github.com/trufflesuite/ganache-core/issues?q=is%3Aopen+is%3Aissue+label%3Aforking), so I'm trying to get it the fixes it needs 😁

**Comment by Bobface** [comment]
Sure, no problem. But before that I think we should wait for @nodar or @amateur-dev first to confirm.

**Comment by RideSolo** [comment]
@Bobface each time you send a transaction the block is auto mined, meaning that the time will increase between the first and second transaction, no matter if you use increate time or not.
maybe you can use "--blockTime Specify blockTime in seconds for automatic mining. If you don't specify this flag, ganache will instantly mine a new block for every transaction. "
when you fork the network specify the block time, and you can put the execution to sleep in between transactions.
another solution will be to stop mining as shown [here](https://github.com/ethereum/go-ethereum/wiki/Management-APIs#miner_stop) and advance block manually after x transactions.

**Comment by Bobface** [comment]
@RideSolo When I combined forking the mainnet with blocktime set to 1 second (`-b 1`) ganache completely glitched out and the tests were stuck in a `getTransactionReceipt` loop during `beforeEach`. I am not sure if combining `-f` with `-b` is supposed to work.

**Comment by RideSolo** [comment]
@Bobface if you tried it then there is a bug, but the second solution worked for me before but I didn't try it with a fork.

**Comment by Solexplorer** [comment]
Hey @amateur-dev,

I don't want to be annoying but can you review the PR or get someone to review it for you? I think @Bobface is open to any feedback you might have.

**Comment by owocki** [comment]
hey @nodar can you pls pay out the work that the dev has done on this so far?  if the scope is complete the bounty should be paid. i know there is an issue with the upstream PR merge.. but we can't guarantee that upstream manintainers are going to merge PRs.

**Comment by amateur-dev** [comment]
Hey Guys, so long story short, the test scripts that have been submitted are sadly not covering the full scope and as we know it. I, definitely, am more than happy to pay to Bob for the efforts put in.  Thank you @Bobface, but sadly, our issue is still not resolved.  Though, we will be working on it.  Just my 2 cents.  #withAllDueRespect

**Comment by Bobface** [comment]
@amateur-dev Could you explain which parts of the full scope the tests are not covering? I have put in comments in the source code listing the parts in the tests for all requirements mentioned in the bounty.

**Comment by amateur-dev** [comment]
Hi @Bobface, oh I meant the "successfully running test".  I did go through your code no issues over there.  I do note that it was or is not part of the scope to deep dive why the tests are failing.  I am more on that part as of now.

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__The funding of 400.0 DAI (400.0 USD @ $1.0/DAI)  attached to this issue has been approved & issued to @bobface.__  

 
* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/amateur-dev/Unipool/1/4143)
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $125,687.80 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)
